### PR TITLE
added LitecoinBlockHeaderAttestation

### PR DESCRIPTION
**Pull request by cloutier**
Litecoin is a bit annoying because it doesn't have RBF yet, but it is very useful for getting very granular timestamps. 

I already have a forked (and _very_ hacky) version of opentimestamps-server that works with Litecoin, and I'm working right now on making those change properly and proposing them upstream. 

It's very likely I'm going to have the budget (and the need) to run a public Litecoin calendar that has a transaction in every block, and I'd much prefer it to be part of the OTS standard.

Any thoughts?

### CODE REVIEW

**Comment by petertodd** [comment]
Concept ACK; annoying that RBF is a barrier... Though it's quite possible that RBF-supporting Litecoin miners do exist.

I'll take a look at this later next week; traveling right now.

**Comment by cloutier** [comment]
I am not to worried about RBF, litecoin core devs seems to think RBF is more suited for a chain with full blocks, so they are planning to enable it for all nodes not to far in the future.

So right now my hacky way of doing it is setting `--btc-min-relay-feerate` high enough to be relayed with a very low`--btc-max-fee`, effectively disabling RBF. I don't think it will be worth it to change anything else in the code to support it.

**Comment by petertodd** [comment]
So a slightly problematic thing about this patch is `verify_against_blockheader()`: what library is that meant to actually work with?

Also, I'd suggest that you remove the comment saying that you could have a more granular proof of existence via Litecoin, because that's not clearly true: the accuracy of Litecoin block timestamps isn't any different than Bitcoin, and trusting them to within a few hours (or even a day) is problematic as miners can get away with lying about them.

**Comment by cloutier** [comment]
I've made it work using a slightly modified version of your python-bitcoinlib, changing only a bit of code about adresses prefixes. I'm still not sure if it would be better to to merge those changes or to do a friendly fork of both python-bitcoinlib and opentimestamp-server into a litestamp project that tracks upstream closely, what do you think?

I updated the wording to be more about how fast you can get your confirmation on Litecoin.

Slightly tangential to this, do you think we could fix this by using something like with the averages of the timestamps of the 1000 block before and after the attestation?

File: opentimestamps/core/notary.py [hunk 1]
```diff
@@ -306,9 +306,9 @@ class LitecoinBlockHeaderAttestation(TimeAttestation):
     of an attestation on OpenTimestamps is always zero.
 
     However, a highly motivated calendar server could have an attestion transaction 
-    in every block every 2.5 minutes on Litecoin, which provides a more
-    granular proof of existence than Bitcoin. A user could have a highly secure 
-    attestation the bitcoin blockchain and a precise one on an altchain such as Litecoin.
+    in every block every 2.5 minutes on Litecoin, which provides a faster confirmation
+    than Bitcoin. A user could fetch a highly secure attestation the bitcoin blockchain 
+    and immediately get one on an altchain such as Litecoin.
```
Comment (petertodd) [inline opentimestamps/core/notary.py]: Heck, lets just delete these two paragraphs for now; no sense getting into the politics of what is or isn't secure in this kind of documentation for now.

**Comment by petertodd** [comment]
> I've made it work using a slightly modified version of your python-bitcoinlib, changing only a bit of code about adresses prefixes. I'm still not sure if it would be better to to merge those changes or to do a friendly fork of both python-bitcoinlib and opentimestamp-server into a litestamp project that tracks upstream closely, what do you think?

Right, so until a supported python-litecoinlib exists, it's better that we leave off the actual validation code and just define the type of proof instead; users of it can validate that for themselves.

> Slightly tangential to this, do you think we could fix this by using something like with the averages of the timestamps of the 1000 block before and after the attestation?

tl;dr: No - interpreting PoW timestamps is really complex... :) There's some posts on the ots-dev mailing list about this, and it's an open problem.

**Comment by cloutier** [comment]
>>  Slightly tangential to this, do you think we could fix this by using something like with the averages of the timestamps of the 1000 block before and after the attestation?

 > tl;dr: No - interpreting PoW timestamps is really complex... :) There's some posts on the ots-dev mailing list about this, and it's an open problem.

For potential future reader, [this is the email](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2016-September/013120.html).

This is quite interesting, do you know if there is any research on the corrolation of unix time/mining hardware ownership between different chains?

> Right, so until a supported python-litecoinlib exists, it's better that we leave off the actual validation code and just define the type of proof instead; users of it can validate that for themselves.

I'm okay with that, I mostly don't want someone the implement a litecoin implementation with a different tag and have conflicting implementations. Let's leave this open until I have my implementation live and done.

Also as opentimestamp will become more used, lots of people will open PR to have their chain in the standard. It would be good to define the minimum requirements for that before the inflow, I was thinking something like: 

- only chain with the most hashrate for a given mining function will be merged in (no bitcoin cash and no dogecoin)
- no chains with a consistant majority miner
- needs a working implementation of an calendar server

What do you think?

**Comment by cloutier** [comment]
I've also removed the comment, I'm going to do some research and come back with an updated one when I can prove my claims.

**Comment by petertodd** [comment]
Re: requirements, I think you can mostly do it via market cap: the threat model of all PoW chains is relatively similar, so there isn't a good argument for having a lot of them given that the security each additional coin could possibly give isn't much.

**Comment by cloutier** [comment]
Setting aside the marketing argument (we timestamp to more chains than any of our competitors!), I think you're right that it doesn't really give additional security or any upside really to the client's proof.

I think it's more about reducing negative externalities, we can push the use cases that needs a complete proof ASAP towards Litecoin, where the block space is much less scarce, and clients can have a proof transaction in every block without causing harm.

**Comment by petertodd** [comment]
I don't think externalities are a good argument; OTS is so scalable the load is a rounding error on any system. Also, it's unclear there are any cases where having a complete proof ASAP on a PoW system makes sense.

**Comment by cloutier** [comment]
bylls.com are doing signed and timestamped price quotes, and ideally they want their users to have a trustless proof before their quote expire after 15 minutes, which leaves only Ethereum Classic and Litecoin as acceptable chains for them. 

I fully admit I might be biased because I work in the same building and have lunch with them every day.

**Comment by petertodd** [comment]
The problem is a trustless timestamp proof with 15 minute resolution just isn't possible, because miners can manipulate timestamps far more than 15 minutes in either direction. For that kind of need we're stuck with trusted timestamps, which IMO is fine given how we can audit them effectively.

**Comment by cloutier** [comment]
The actual time in this usecase is irrelevant. Only the chronological order of the timestamps is important to them. They only want to allow users to prove their receipts are valid between the time bylls' key was created/timestamped and the time that their key's revocation was timestamped/published far and wide.

There could be an attack vector where one timestamps a revocation in every block, and only publish the the most convinient proof after the facts. To avoid this, they are thinking about publishing a revocation straight in a OP_RETURN. Doing this on Litecoin instead of Bitcoin would have much less of a negative impact.

I've updated my code so that it would only display the LTC proof correctly and raise an NotImplementedError when trying to verify. I could update it when there is a good "python-litecoinlib" (potentially made by me).

My number one goal with this PR is making sure that someone else working on Litecoin would end up with an implementation compatible with mine, and not fragment the community. I think it's also a good way to get up to speed with how the project works, and thanks for the lengthy discussions about what is a good or bad idea.

**Comment by petertodd** [comment]
Ah, so cronological ordering is a subtle thing, and OpenTimestamps is probably not what you want there.

In fact, mind if we take that topic to the ots-dev mailing list? This is a topic of interest to the community, and I think it'd be worth having a thread about it outside of github. Specifically, I'd like to know what exactly you need to prove here.

**Comment by RCasatta** [comment]
Concept ACK

**Comment by petertodd** [comment]
Merging, because we have ETH, and this is _way_ less sketchy. :)

**Comment by cloutier** [comment]
Thanks for merging!

I'm collecting more info from the bylls team (an useful exercices in and of itself) and I'm drafting a post for the mailing list with their exact requirements and what I'm considering implementing to help them.

**Comment by petertodd** [comment]
Sounds good, thanks!
